### DISCUSSION TITLE
Installation woes

### DISCUSSION THREAD

**Discussion by bwlove319** [theory/sqitch#609] [General]
TL;DR = This is far too difficult to install. What steps can you take to simplify?

This is a good product overall.  But the barrier to entry is very high.  Sqitch is written in an old language that is difficult to maintain, and installing it requires at least one additional SDK, package manager, and supporting libs because the pieces it requires must be built/compiled on the local machine.
In this day and age, where it could be single Python module or something equally simple, people will rarely choose to embrace apps with long, arduous, fragile installation procedures unless they absolutely have to.
It is literally taking me 10 times longer to install and run sqitch, than it took me to create the extensive SQL scripts.
Other people are experiencing the same pain, as evidenced here:
https://github.com/sqitchers/docker-sqitch/issues/26

Can you please update your documentation to clarify some things:
- for Snowflake, why would we need to install the ODBC driver AND specify where the snowsql client is?  You talk about both, but they should be mutually exclusive.
- How do you install an ODBC driver manager on modern versions of Linux that no longer natively support it.  I understand we can Google that, but the results are convoluted and difficult to decipher.  You should either publish a set of steps or point to another site that does.  

It's impractical to use any tool in a CI/CD pipeline that needs 20+ installation steps just to be executable.  `apt install sqitch` or `yum install sqitch` should be all that is required... not 5 hours of research, setup, downloading, and compiling.  If you want a broader userbase, your app needs to be usable by people who don't have a PhD in Perl.

**Comment by bwlove319** [comment]
So I'm about 4 hours into trying to install this.  
Odbc works fine:
![image](https://user-images.githubusercontent.com/87549506/156461109-bb76b851-356c-42a5-afe1-369bdcab2d0e.png)

But sqitch doesn't:
![image](https://user-images.githubusercontent.com/87549506/156461238-33b4b9d8-1c49-4176-bd25-36e3b6d5b7df.png)

Executing this command:
`cpan install DBD::ODBC`
installed version 1.61, which apparently is not sufficient...?  Apparently sqitch requires 1.59 and ONLY 1.59, which is odd.  Since there's apparently no way to tell that prior to executing sqitch.

So the installation flow is:
1. install odbc
2. install sqitch
3. execute sqitch to find out what other modules need to be installed that aren't mentioned in the installation docs anywhere
4. attempt to download and compile those modules
5. download and install all the pieces required in order to build the modules sqitch needs
6. download and compile the modules sqitch needs
7. execute sqitch to find out which versions from step 6 are wrong
8. download and install correct versions of sqitch dependencies
9. repeat steps 3..9 until sqitch executes without error.

And we're supposed to embrace this as a portable, automatable DB change management tool?  How are we supposed to put all those steps into a buildspec for AWS, or a Jenkinsfile?

There must be a better way.

**Reply by theory** [reply]
I suspect you're installed DBD::ODBC against a different version of Perl than you installed Sqitch against. Compare

```
which perl
```

to

```
head -1 $(which sqitch) 
```

**Reply by bwlove319** [reply]
> I suspect you're installed DBD::ODBC against a different version of Perl than you installed Sqitch against. Compare
> 
> ```
> which perl
> ```
> 
> to
> 
> ```
> head -1 $(which sqitch) 
> ```

As it turns out, the commands I had were accurate, but apparently the order you execute them in matters.
I wiped out my devcontainer and recreated it.  Then after executing the installation commands I'd collected in a different order it did work.  No idea why, but I'm not Perl savvy.
And the 1.61 version of DBD::ODBC I'd installed worked fine after that.

For reference, here's my command list:


![image](https://user-images.githubusercontent.com/87549506/156629478-2390205a-540b-402c-8519-78166b617e0d.png)


I was initially missing unixodbc-dev, which took me a while to figure out.
